### ISSUE TITLE
Zeebe v1.1.1 cannot create snapshots on Windows OS

### ISSUE THREAD

**Issue by sergeylebed** [camunda-cloud/zeebe#7570]
**Describe the bug**
Zeebe v1.1.1 cannot create snapshots on Windows OS

**To Reproduce**
Get the default distribution package and run it with default settings (1 partition).
Provide a test load (e.g. run a canary workflow every minute).
For easy reproduction you can set snapshotting period to 1m.

**Expected behavior**
Snapshots are created.

**Log/Stacktrace**
<details>
<p>

```
My logs:
2021-07-30 16:11:19.838 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-0] INFO 
      io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 1477045 is committed, current commit position is 1477045. After that snapshot will be committed.
2021-07-30 16:11:19.846 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-1] ERROR
      io.camunda.zeebe.logstreams.snapshot - Unexpected exception occurred on moving valid snapshot.
java.io.UncheckedIOException: java.nio.file.AccessDeniedException: C:\Dev\Zeebe.io\camunda-cloud-zeebe-1.1.1\data\raft-partition\partitions\1\snapshots\965847-2-1477029-9223372036854775807
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.moveToSnapshotDirectory(FileBasedSnapshotStore.java:477) ~[zeebe-snapshots-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.newSnapshot(FileBasedSnapshotStore.java:409) ~[zeebe-snapshots-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.persistInternal(FileBasedTransientSnapshot.java:133) ~[zeebe-snapshots-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.lambda$persist$3(FileBasedTransientSnapshot.java:101) ~[zeebe-snapshots-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorControl.lambda$call$0(ActorControl.java:136) ~[zeebe-util-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorJob.invoke(ActorJob.java:62) [zeebe-util-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [zeebe-util-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:94) [zeebe-util-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [zeebe-util-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.util.sched.ActorThread.run(ActorThread.java:191) [zeebe-util-1.1.1.jar:1.1.1]
Caused by: java.nio.file.AccessDeniedException: C:\Dev\Zeebe.io\camunda-cloud-zeebe-1.1.1\data\raft-partition\partitions\1\snapshots\965847-2-1477029-9223372036854775807
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:89) ~[?:?]
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103) ~[?:?]
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108) ~[?:?]
	at sun.nio.fs.WindowsFileSystemProvider.newFileChannel(WindowsFileSystemProvider.java:120) ~[?:?]
	at java.nio.channels.FileChannel.open(FileChannel.java:292) ~[?:?]
	at java.nio.channels.FileChannel.open(FileChannel.java:345) ~[?:?]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.tryAtomicDirectoryMove(FileBasedSnapshotStore.java:511) ~[zeebe-snapshots-1.1.1.jar:1.1.1]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.moveToSnapshotDirectory(FileBasedSnapshotStore.java:468) ~[zeebe-snapshots-1.1.1.jar:1.1.1]
	... 10 more
```

```
These logs from Svetlana
io.camunda.zeebe.logstreams.snapshot - Unexpected exception occurred on moving valid snapshot.
java.io.UncheckedIOException: java.nio.file.AccessDeniedException: C:\programs\camunda-cloud-zeebe-1.1.0\data\raft-partition\partitions\1\snapshots\592-1-1961-1964
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.moveToSnapshotDirectory(FileBasedSnapshotStore.java:477) ~[zeebe-snapshots-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.newSnapshot(FileBasedSnapshotStore.java:409) ~[zeebe-snapshots-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.persistInternal(FileBasedTransientSnapshot.java:133) ~[zeebe-snapshots-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.lambda$persist$3(FileBasedTransientSnapshot.java:101) ~[zeebe-snapshots-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorControl.lambda$call$0(ActorControl.java:136) ~[zeebe-util-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorJob.invoke(ActorJob.java:62) [zeebe-util-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [zeebe-util-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:94) [zeebe-util-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [zeebe-util-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.util.sched.ActorThread.run(ActorThread.java:191) [zeebe-util-1.1.0.jar:1.1.0]
Caused by: java.nio.file.AccessDeniedException: C:\programs\camunda-cloud-zeebe-1.1.0\data\raft-partition\partitions\1\snapshots\592-1-1961-1964
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:89) ~[?:?]
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103) ~[?:?]
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108) ~[?:?]
	at sun.nio.fs.WindowsFileSystemProvider.newFileChannel(WindowsFileSystemProvider.java:116) ~[?:?]
	at java.nio.channels.FileChannel.open(FileChannel.java:292) ~[?:?]
	at java.nio.channels.FileChannel.open(FileChannel.java:345) ~[?:?]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.tryAtomicDirectoryMove(FileBasedSnapshotStore.java:511) ~[zeebe-snapshots-1.1.0.jar:1.1.0]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.moveToSnapshotDirectory(FileBasedSnapshotStore.java:468) ~[zeebe-snapshots-1.1.0.jar:1.1.0]
	... 10 more
```
</p>
<p>
From the process monitor I see that files are created in the snapshot folder (by renaming pending/<id>) and then deleted. I don't see any file related errors in the process monitor, so I'm might be some logical error in the upper libraries rather than an actual access deny error on file system level.
<p>
<p>
I've checked version 0.26.4. It creates snapshots normally under the same condition.
</p>

</details>

**Environment:**
 - OS: Windows OS (Windows Server 2016 Standard)
 - Zeebe Version: 1.1.1
 - Configuration: default from the distribution package.

**Comment by Zelldon** [comment]
@npepinpe @sdorokhova was able to reproduce it

**Comment by npepinpe** [comment]
It's possibly related to the fsync dance we added for Linux - when moving/renaming an entry in the filesystem, you need to move it and then fsync it's new parent directory to ensure that it will be visible on crash recovery. In Java there's no easy way to get a  file descriptor and fsync it, so you do something like:

```java
FileChannel.open(path).force(true);
```

This works even for directory (though actually using the file channel for writing or reading would produce errors). We use this trick right now for directories, but don't specify any permissions to avoid confusion (i.e. don't accidentally think you can read and/or write). I suspect this may not work on Windows, however, from bug reports in the Lucene project. Their solution was to ignore errors on fsync of directories if you're on Windows. See IOUtils.java from their project:

https://github.com/apache/lucene/blob/645b64ef4e30baf18a47cba3110e663242c092df/lucene/core/src/java/org/apache/lucene/util/IOUtils.java#L458-L487

I'm not super keen on introducing OS specific switches however, since we explicitly only support Docker and Kubernetes as runtime platform. I'm open to investigating different ways of fsync'ing a directory, but that'll be up to those running on Windows, since I don't have access to a Windows machine. You could try other ways of obtaining a file descriptor and syncing it, e.g.:

```java
new RandomAccessFile(new File("myDirectory")).getFD().sync();
new FileOutputStream(new File("myDirectory")).getFD().sync();
```

I suspect these wouldn't work either, but who knows.

**Comment by npepinpe** [comment]
It seems the only way to get a file descriptor for a directory in Windows is to use the flag `FILE_FLAG_BACKUP_SEMANTICS`, which isn't directly available via the JDK. I'd rather add a branch to sync/not sync than write Windows specific JNI code for this :smile:

**Comment by npepinpe** [comment]
@menski - once again, how important is the Windows use case?

**Comment by npepinpe** [comment]
This is fixed in 1.1.3 and 1.2.0.
